File: 2022/kotlin/src/main/kotlin/Day4.kt
import java.io.BufferedReader
import java.io.FileInputStream
import java.io.InputStreamReader

fun main() {
    getQuantityOfOverlappedPairs()
}

private fun getQuantityOfOverlappedPairs() {
    var countAllOverlapping = 0
    var countRangesOverlapping = 0
    BufferedReader(InputStreamReader(FileInputStream("src/main/resources/day4.txt"))).use { br ->
        var line: String?
        while (br.readLine().also { line = it } != null) {
            val sections =
                line!!.split(",".toRegex()).dropLastWhile { it.isEmpty() }
                    .toTypedArray()
            val firstSectionRange: Set<Int> = getSetRangefromSection(sections[0])
            val secondSectionRange: Set<Int> = getSetRangefromSection(sections[1])
            if (allSectionOverlaps(firstSectionRange, secondSectionRange)) {
                countAllOverlapping++
            }
            if (rangeSectionOverlaps(firstSectionRange, secondSectionRange)) {
                countRangesOverlapping++
            }
        }
    }
    println(countAllOverlapping)
    println(countRangesOverlapping)
}

private fun getSetRangefromSection(section: String): Set<Int> {
    val numbers = section.split("-".toRegex()).dropLastWhile { it.isEmpty() }
        .toTypedArray()
    val start = numbers[0].toInt()
    val end = numbers[1].toInt()
    val result: MutableSet<Int> = HashSet()
    for (i in start..end) {
        result.add(i)
    }
    return result
}

private fun allSectionOverlaps(first: Set<Int>, second: Set<Int>): Boolean {
    return first.containsAll(second) || second.containsAll(first)
}

private fun rangeSectionOverlaps(first: Set<Int>, second: Set<Int>): Boolean {
    for (value in first) {
        if (second.contains(value)) return true
    }
    return false
}

File: 2022/javascript/day4.js
const range = (start, stop, step) => Array.from({ length: (stop - start) / step + 1 }, (_, i) => start + i * step);

const lines = require("fs").readFileSync("resources/day4.txt", "utf-8").split("\r\n");

const getSetFromRangeSection = (section) => {
  const [start, end] = section.split("-");
  return range(Number(start), Number(end), 1);
};

const doesSetOverlaps = (first, second) => first.every((e) => second.includes(e)) || second.every((e) => first.includes(e));

const doesRangeOverlaps = (first = [], second) => first.some((e) => second.includes(e));

var countAllOverlaps = 0;
var countRangeOverlaps = 0;

const calculateSetRanges = () => {
  for (const line of lines) {
    const [firstSection, secondSection] = line.split(",");
    const firstSectionRange = getSetFromRangeSection(firstSection);
    const secondSectionRange = getSetFromRangeSection(secondSection);
    if (doesSetOverlaps(firstSectionRange, secondSectionRange)) {
      countAllOverlaps++;
    }
    if (doesRangeOverlaps(firstSectionRange, secondSectionRange)) {
      countRangeOverlaps++;
    }
  }
  console.log(countAllOverlaps);
  console.log(countRangeOverlaps);
};

calculateSetRanges();
